geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/KryoLazyDensityFilter.java

```diff
@@ -0,0 +1,359 @@
+            }
+        }
+    }
+
+    /**
+     * Evaluates an arbitrary expression against the simple feature to return a weight
+     */
+    private double getWeightFromExpression(Expression e, SimpleFeature sf) {
+        Double d = e.evaluate(sf, Double.class);
+        if (d == null) {
+            return 0.0;
+        } else {
+            return d;
+        }
+    }
+
+    /*
+     * Returns the weight of a simple feature
+     */
+    double weightFn(SimpleFeature sf) {
+        double returnVal = 0.0;
+        if (weightIndex == -2) {
+            returnVal = 1.0;
+        } else if (weightIndex == -1) {
+            try {
+                Expression expression = ECQL.toExpression((String) options.get(WEIGHT_OPT));
+                returnVal = getWeightFromExpression(expression, sf);
+            } catch (Exception e) {
+                System.out.println("Exception");
+            }
+        } else if (sft.getDescriptor(weightIndex).getType().getBinding() == Double.class) {
+            returnVal = getWeightFromDouble(weightIndex, sf);
+        } else {
+            returnVal = getWeightFromNonDouble(weightIndex, sf);
+        }
+        return returnVal;
+    }
+
+    boolean isPoints() {
+        GeometryDescriptor gd = sft.getGeometryDescriptor();
+        return (gd != null && gd.getType().getBinding() == Point.class);
+    }
+
+    Map.Entry<Tuple2<Integer, Integer>, Double> writePoint(SimpleFeature sf, Double weight) {
+        return writePointToResult((Point) sf.getAttribute(geomIndex), weight);
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writeNonPoint(Geometry geom, Double weight) {
+        return writePointToResult(safeCentroid(geom), weight);
+    }
+
+    private Point safeCentroid(Geometry geom) {
+        Polygon p = geom.noPolygon();
+        Point centroid = p.getCentroid();
+        if (Double.isNaN(centroid.getCoordinate().x) || Double.isNaN(centroid.getCoordinate().y)) {
+            return p.getEnvelope().getCentroid();
+        } else {
+            return centroid;
+        }
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writePointToResult(Point pt, Double weight) {
+        return writeSnappedPoint(gridSnap.i(pt.getX()), gridSnap.j(pt.getY()), weight);
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writeSnappedPoint(int x, int y, Double weight) {
+        Tuple2 tuple = new Tuple2(x, y);
+        return new AbstractMap.SimpleEntry(tuple, weight);
+
+    }
+
+    private void writePointToResult(Coordinate pt, Double weight) {
+        writeSnappedPoint(gridSnap.i(pt.x), gridSnap.j(pt.y), weight);
+    }
+
+    private void writePointToResult(double x, double y, double weight) {
+        writeSnappedPoint(gridSnap.i(x), gridSnap.j(y), weight);
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell v) throws IOException {
+        byte[] encodedSF = CellUtil.cloneValue(v);
+        SimpleFeature sf = serializer.deserialize(encodedSF);
```

Comment:
no point in deserializing if we're not checking the result... also if we are deserializing see earlier comment about not cloning the cell value